src/lab16.py

```diff
@@ -221,28 +221,35 @@ def __init__(self, node, parent, previous):
         self.width = None
         self.height = None
 
-        self.dirty_width = True
-        self.dirty_height = True
-        self.dirty_x = True
-        self.dirty_y = True
+        self.zoom_field = DependentField(self, "zoom")
+        self.x_field = DependentField(self, "x")
+        self.y_field = DependentField(self, "y")
+        self.width_field = DependentField(self, "width")
+        self.height_field = DependentField(self, "height")
 
     def layout(self):
-        self.zoom = self.parent.zoom
-        self.width = self.parent.width
-        self.x = self.parent.x
+        parent_zoom = self.zoom_field.read(self.parent.zoom_field)
+        self.zoom = self.zoom_field.set(parent_zoom)
+
+        parent_width = self.width_field.read(self.parent.width_field)
+        self.width = self.width_field.set(parent_width)
+
+        parent_x = self.x_field.read(self.parent.x_field)
+        self.x = self.x_field.set(parent_x)
 
         if self.previous:
-            self.y = self.previous.y + self.previous.height
+            prev_y = self.y_field.read(self.previous.y_field)
+            prev_height = self.y_field.read(self.previous.height_field)
+            self.y = self.y_field.set(prev_y + prev_height)
         else:
-            self.y = self.parent.y
+            parent_y = self.y_field.read(self.parent.y_field)
+            self.y = self.y_field.set(parent_y)
 
         for word in self.children:
             word.layout()
 
         if not self.children:
-            self.height = 0
-            self.parent.dirty_height = True
-            mark_dirty(self.parent)
+            self.height = self.height_field.set(0)
```

Comment:
This misses dirtying self.parent.

Reply:
The parent should depend on the child height and be dirtied by the set operation.